Add process execution simulation to test harness

weiiwang01 opened the pull request:
The Pebble process execution plays an important role in a lot of charms. However, the ops framework testing module currently lacks built-in functionality to simulate this process, leading many charms to roll their own patching and mocking mechanisms for unit tests.

To address this problem, this pull request introduces `ops.testing.Harness.handle_exec`, a new feature that enables the simulation of Pebble process execution within the Ops framework testing module.

See [spec OP039](https://docs.google.com/document/d/1V7nzFzSOIMVlPF-HCvTZZyWOGlHnCS__RyL5352P2e4/edit) for discussion.

benhoyt commented:
@weiiwang01 Is this ready for review? If so, I can take a look tomorrow.

Thanks for the `pebble.py` None-safety fix. Are you able to separate that out into a separate PR, with a regression test of that code path if possible?

One other thing: it looks like the docs are failing right now. You can generate those locally with `tox -e docs` and then `open docs/_build/html/index.html` to review the docs locally.

weiiwang01 commented:
> @weiiwang01 Is this ready for review? If so, I can take a look tomorrow.
> 
> Thanks for the `pebble.py` None-safety fix. Are you able to separate that out into a separate PR, with a regression test of that code path if possible?
> 
> One other thing: it looks like the docs are failing right now. You can generate those locally with `tox -e docs` and then `open docs/_build/html/index.html` to review the docs locally.

Yes, this is ready for review. I have fixed the document issue and reverted changes in `pebble.py`. I will create a new PR for the None safety issue later.

benhoyt commented on test/test_testing.py:
```diff
@@ -4841,3 +4842,270 @@ def test_errors(self):
+        self.assertEqual(args_history[-1].group_id, 4)
+        self.assertDictEqual(args_history[-1].environment, {"foo": "hello", "foobar": "barfoo"})
+
+    def test_registration_order(self):
```
What's actually being tested here? That length matters and not registration order? Can we do the same thing without inspecting implementation details, for example just by doing two registrations in one order and then in the opposite order and seeing that the matching is correct?

weiiwang01 replied:
See this reply https://github.com/canonical/operator/pull/993#discussion_r1302970333

benhoyt replied:
Let's remove this test once we're using a dict for the registered handlers (or simplify it per my suggestion above in this thread).

weiiwang01 replied:
Removed, thanks!

benhoyt reviewed (changes requested):
This is great work again, thanks @weiiwang01. I've left a bunch of comments, but most of them are minor -- structurally it seems fine.

benhoyt commented on ops/testing.py:
```diff
@@ -106,6 +109,41 @@
+    """
+    exit_code: int = 0
+    stdout: Union[str, bytes] = b""
+    stderr: Union[str, bytes] = b""
```
Makes sense, thanks for the explanation.

benhoyt commented on ops/testing.py:
```diff
@@ -1544,6 +1583,93 @@ def evaluate_status(self) -> None:
+        if (handler is None and result is None) or (handler is not None and result is not None):
+            raise TypeError("Either handler or result must be provided, but not both.")
+        container_name = container if isinstance(container, str) else container.name
+        if isinstance(result, int):
+            result = ExecResult(exit_code=result)
+        if isinstance(result, (str, bytes)):
+            result = ExecResult(stdout=result)
```
How about this, to help people not using static type checking? The "not bool" test is because bools are ints, but we don't want someone confusing things by returning True (1) or False (0).

```suggestion
        if isinstance(result, int) and not isinstance(result, bool):
            result = ExecResult(exit_code=result)
        elif isinstance(result, (str, bytes)):
            result = ExecResult(stdout=result)
        elif not isinstance(result, ExecResult):
            raise TypeError(f"result must be int, str, bytes, or ExecResult, not {result.__class__.__name__}")
```

weiiwang01 replied:
Never know the `bool` is `int`, but I guess it makes sense, updated, thanks!

benhoyt commented on ops/testing.py:
```diff
@@ -2321,6 +2447,53 @@ def _check_protocol_and_port(self, protocol: str, port: Optional[int]):
+    def wait(self):
+        if self._is_timeout:
+            raise pebble.TimeoutError(
+                f'timed out waiting for change 123 ({self._timeout} seconds)'
```
What I was suggesting here is that we simply remove the change ID, the " 123" part, from the error message entirely. I think it will be confusing to have a hard-coded value here.

weiiwang01 replied:
Gotcha, removed, thanks!

benhoyt commented on ops/testing.py:
```diff
@@ -2321,6 +2440,49 @@ def _check_protocol_and_port(self, protocol: str, port: Optional[int]):
+        if self._exit_code != 0:
+            raise pebble.ExecError(self._command, cast(int, self._exit_code), None, None)
+
+    def wait_output(self) -> Tuple[Optional[AnyStr], Optional[AnyStr]]:
```
Perhaps it has no effect, but let's change it to the same signature as the real `wait_output` in this PR to avoid confusion.

weiiwang01 replied:
Updated, thanks!

benhoyt commented on ops/testing.py:
```diff
@@ -2337,6 +2499,23 @@ def __init__(self, backend: _TestingModelBackend, container_root: pathlib.Path):
+                self._exec_handlers.insert(idx, (prefix, handler))
+                inserted = True
+        if not inserted:
+            self._exec_handlers.append((prefix, handler))
```
Yeah, I think we should -- the current code is fairly tricky and complex, for no practical benefit in this use case. We're only ever going to have a small handful of registered handlers, so whizzing through them linearly each time is fine. To get simple code, let's use a dict and a `_find_exec_handler` along the lines of what I suggested.

weiiwang01 replied:
Okay, updated, I tweaked the lookup process a little bit try to make it even more simpler, thanks!

weiiwang01 replied:
Yes, that's more clearer, updated, thanks!

benhoyt reviewed (changes requested):
Thanks for the updates! This is very close now. Let's just make the simplification to use a dict for the registered handlers, and I've added a couple of other minor suggestions.

I also tweaked the handle_exec docstring and fixed it in a couple of places, and reverted the line-break change to ops/pebble.py to avoid that in the diff.

benhoyt commented on ops/testing.py:
```diff
@@ -2888,10 +2881,10 @@ def remove_path(self, path: str, *, recursive: bool = False):
-            if tuple(command[:len(command_prefix)]) == command_prefix:
-                return handler
-        return None
+        for prefix_len in reversed(range(len(command) + 1)):
```
Clever and clear approach to look it up this way -- I like it!

weiiwang01 replied:
Thanks!

benhoyt reviewed (approved):
Excellent, thanks. Love the new simple lookup technique! Just a couple of nits and then we're good to go.

weiiwang01 commented on ops/testing.py:
```diff
@@ -1544,6 +1582,87 @@ def evaluate_status(self) -> None:
+          :class:`ops.testing.ExecResult` as the simulated process outcome. For simpler cases, the
+          handler can return a string (or byte string)
+          (equivalent to ``ExecResult(exit_code=0, stdout=str_or_bytes)``) or an integer
+          (equivalent to ``ExecResult(exit_code=return_value)``).
```
Understood, updated. However, I can't change `handler = lambda _: result` because of the linting checks.

weiiwang01 commented on ops/testing.py:
```diff
@@ -1544,6 +1583,97 @@ def evaluate_status(self) -> None:
+            raise TypeError("Either handler or result must be provided, but not both.")
+        container_name = container if isinstance(container, str) else container.name
+        if result is None:
+            pass
```
Understood, updated. However, I can't change `handler = lambda _: result` because of the linting checks.

benhoyt replied:
Ugh, I hate it when our tools get in the way of clean simple code. I guess we could do this, but I'm not sure it's any better. I leave it up to you. I'll probably merge this tomorrow!

```python
            # autopep8: off
            handler = lambda _: result  # noqa
            # autopep8: on
```

weiiwang01 replied:
Three annotations for a single statement might be a little bit too much. I'll probably keep the original statement, thanks.

weiiwang01 commented on ops/testing.py:
```diff
@@ -2888,10 +2881,10 @@ def remove_path(self, path: str, *, recursive: bool = False):
+        for prefix_len in reversed(range(len(command) + 1)):
+            command_prefix = tuple(command[:prefix_len])
+            if command_prefix in self._exec_handlers:
+                return self._exec_handlers[command_prefix]
```
Yes, it's more clearer, updated, thanks!

benhoyt commented:
prometheus-k8s-operator unit tests are failing, but that's not related to this PR -- I've made them aware.

Merging! Thanks for your work on this, @weiiwang01.
